### PR TITLE
Refactor and harmonize CSV and SQL data sources

### LINKED ISSUE #716: Refactor CSV dataSource
Create grid elements from a SQL database.

### PR DESCRIPTION
**Pull request by jo-bao**
Resolves #716 
- refactor CSV and SQL data sources to use the methods for creating entities
- implement IdCoordinateSource for SQL

### CODE REVIEW

**Comment by jo-bao** [comment]
!test

**Comment by jo-bao** [comment]
!test

**Comment by jo-bao** [comment]
!test

**Comment by t-ober** [comment]
@sonatype-lift test

**Comment by t-ober** [comment]
@sonatype-lift help

File: src/main/java/edu/ie3/datamodel/io/source/IdCoordinateSource.java [hunk 1]
```diff
@@ -5,48 +5,86 @@
...
+import org.apache.commons.lang3.tuple.Pair;
 import org.locationtech.jts.geom.Point;
 
 /**
  * This class serves mapping purposes between the ID of a coordinate and the actual coordinate with
  * latitude and longitude values, which is especially needed for data source that don't offer
  * combined primary or foreign keys.
  */
-public interface IdCoordinateSource extends DataSource {
+public class IdCoordinateSource implements DataSource {
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/IdCoordinateSource.java:23]: For weather coordinates, we have some relevant changes in #694, something we should keep in mind here. I think the changes can be combined though.

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
Nice work, looks well thought-out for the most part. I got some points below, I think they're more than enough for now :D
Please hit me up with any questions

File: src/main/java/edu/ie3/datamodel/io/source/sql/SqlTimeSeriesSource.java [hunk 2]
```diff
@@ -18,28 +18,85 @@
...
-public class SqlTimeSeriesSource<V extends Value> extends SqlDataSource<TimeBasedValue<V>>
-    implements TimeSeriesSource<V> {
-  private static final String WHERE = " WHERE ";
-  private static final String TIME_SERIES = "time_series";
+public class SqlTimeSeriesSource<V extends Value> extends TimeSeriesSource<V> {
+
+  protected static final Logger log = LoggerFactory.getLogger(SqlTimeSeriesSource.class);
+  private final SqlDataSource dataSource;
 
   private final UUID timeSeriesUuid;
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/java/edu/ie3/datamodel/io/source/sql/SqlTimeSeriesSource.java:32]: Seems to not apply anymore

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy
[the remark this answers is not in the document]
Comment (jo-bao) [reply src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy:1]: That's the plan, but I haven't tested it yet.

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
Ok these are my comments for now, haven't checked the tests yet because I think it makes more sense once this is final.

Two more things to check out besides those comments:
- SonarQube reports 20 new code smells, could you take care of these? Some might even be connected 
- Readthedocs documentation might need some updates, found at least `csvfiles.md` which has outdated info

File: src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java [hunk 3]
```diff
@@ -117,34 +186,128 @@
...
+    }
+    return Stream.empty();
+  }
+
+  /**
+   * Creates a stream with maps representing a data point in the SQL data source using an entity
+   * class.
    */
-  protected List<T> executeQuery(String query, AddParams addParams) {
+  protected Stream<Map<String, String>> buildStreamByEntityClass(
```
[the remark this answers is not in the document]
Comment (jo-bao) [reply src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java:141]: I think we'll need it when we implement the SQL sources for system participants and grid elements.

File: src/main/java/edu/ie3/datamodel/io/source/WeatherSource.java [hunk 4]
```diff
@@ -5,44 +5,151 @@
...
+   * @param a the first time series to merge
+   * @param b the second time series to merge
+   * @return merged time series with a's UUID
+   */
+  protected <V extends Value> IndividualTimeSeries<V> mergeTimeSeries(
+      IndividualTimeSeries<V> a, IndividualTimeSeries<V> b) {
+    SortedSet<TimeBasedValue<V>> entries = a.getEntries();
+    entries.addAll(b.getEntries());
+    return new IndividualTimeSeries<>(a.getUuid(), entries);
+  }
```
[the remark this answers is not in the document]
Comment (jo-bao) [reply src/main/java/edu/ie3/datamodel/io/source/WeatherSource.java:98]: Oh yes, I'll shift it to CsvWeatherSource.

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java [hunk 5]
```diff
@@ -154,7 +152,7 @@ public int getCoordinateCount() {
    *
    * @return Stream with mappings from field identifiers to attributes
    */
-  protected Stream<Map<String, String>> buildStreamWithFieldsToAttributesMap() {
+  public Stream<Map<String, String>> buildStreamWithFieldsToAttributesMap() {
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java:155]: Why?

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java [hunk 6]
```diff
@@ -122,24 +134,28 @@ public Optional<V> getValue(ZonedDateTime time) {
...
       return new IndividualTimeSeries<>(timeSeriesUuid, timeBasedValues);
     } catch (FileNotFoundException e) {
       throw new SourceException("Unable to find a file with path '" + filePath + "'.", e);
     } catch (IOException e) {
       throw new SourceException("Error during reading of file'" + filePath + "'.", e);
     }
   }
+
+  private Set<TimeBasedValue<V>> getTimeBasedValueSet(
+      Function<Map<String, String>, Optional<TimeBasedValue<V>>> fieldToValueFunction,
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java:153]: Isn't `fieldToValueFunction` always the function `createTimeBasedValue`?

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
Almost there, I think, besides some some pedantic points from my side.
Please also take a look at SonarQube, some new issues popped up :)

File: src/main/java/edu/ie3/datamodel/io/source/WeatherSource.java [hunk 7]
```diff
@@ -5,44 +5,131 @@
...
+   * @return merged time series with a's UUID
    */
-  Map<Point, IndividualTimeSeries<WeatherValue>> getWeather(
-      ClosedInterval<ZonedDateTime> timeInterval);
+  protected <V extends Value> IndividualTimeSeries<V> mergeTimeSeries(
+      IndividualTimeSeries<V> a, IndividualTimeSeries<V> b) {
+    SortedSet<TimeBasedValue<V>> entries = a.getEntries();
+    entries.addAll(b.getEntries());
+    return new IndividualTimeSeries<>(a.getUuid(), entries);
+  }
```
[the remark this answers is not in the document]
Comment (jo-bao) [reply src/main/java/edu/ie3/datamodel/io/source/WeatherSource.java:67]: I think you mean CsvWeatherSource :D

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java (hunk 6 above)
Comment (sebastian-peter) [reply src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java]: Ok what you did there was also helpful, it's not what I meant though :D Not that important though

**Review by sebastian-peter — APPROVED** [review]
Thanks for putting all this work in here! I think you've found some really nice abstractions that should make the codebase a lot more usable.

**Pull request merged**